fix: SQL queries to improve Postgres support

c-p-b opened the pull request:
Changes to improve Postgres support

Verequies committed:
fix: Postgres Compatibility for Standard Views

Core/User DocType: Add 'group_by' for PostgreSQL

Signed-off-by: Verequies <hamishclaxton@gmail.com>

fix: Postgres Compatibility for Standard Views

Desk/Dashboard DocType: Fix lowercase table name

Signed-off-by: Verequies <hamishclaxton@gmail.com>

Desk/Page/Setup Wizard: Fix data not being commited to database

Signed-off-by: Verequies <hamishclaxton@gmail.com>

Model/db_query: Fix queries with order_by and group_by for PostgreSQL

Signed-off-by: Verequies <hamishclaxton@gmail.com>

Model/db_query: Fix order_by without table name for PostgreSQL

Signed-off-by: Verequies <hamishclaxton@gmail.com>

c-p-b committed:
fix: Postgres Compatibility

* Handle inconsistencies in type handling in DatabaseQuery & Database
  APIs
* Update incompatible queries with frappe.qb notation
* Fixed use cases discovered by failing ERPNext CI tests

fix: db independent syntax for user_type

fix: handle postgres datetime values

feat: add ability to auto commit on db inserts

feat: add ability to escape underscore in postgres

fix: handle missing data in test runner bootstrapping

fix: db independent syntax for queries

fix: refactor to use qb

fix: update cache for language

fix: use pluck in email_queue

Co-authored-by: gavin <gavin18d@gmail.com>

fix: don't auto insert on tests for make_property_setter

fix: remove auto_commit in custom_field insertion

fix: remove auto_commit functionality

fix: review comments

fix: revert link validation

fix: style suggestion for readability

Co-authored-by: gavin <gavin18d@gmail.com>

fix: revert .lower() in link validation

fix: add rollback for setup_wizard

Revert "fix: add rollback for setup_wizard"

This reverts commit 83b3b0913db17718ccd5edae01858cff15603829.

Revert "feat: add ability to escape underscore in postgres"

This reverts commit 8ed9c2aa3306438e94bb813f60e65b416d0b947b.

fix: more concise representation of order fields

Co-authored-by: gavin <gavin18d@gmail.com>

surajshetty3416 commented:
@cpdeethree Since you are fixing broken queries for Postgres, try and replace broken raw SQLs with [frappe.qb](https://frappeframework.com/docs/v13/user/en/api/query-builder) as much as possible. It will also help in avoiding possible "SQLi"s.

ankush commented:
> @cpdeethree Since you are fixing broken queries for Postgres, try and replace broken raw SQLs with frappe.qb as much as possible. It will also help in avoiding possible "SQLi"s.


@surajshetty3416 there aren't many raw query changes in this PR, we can skip this for now. This PR's aim is to only add changes that are required to make ERPNext installable and tests passable on ERPNext PRs.

surajshetty3416 commented:
> @surajshetty3416 there aren't many raw query changes in this PR, we can skip this for now. This PR's aim is to only add changes that are required to make ERPNext installable and tests passable on ERPNext PRs.

This seems to be the perfect opportunity to replace some `multisql` in the codebase. There are only a few instances of them it shouldn't take much time to do it... also, [adding more of such code](https://github.com/frappe/frappe/pull/15354/files#diff-9c7814ddc80a02f96e8b9c1586b663a81e66d5f53f802429142e8aaafaacc122R494) does not make sense.

Since the nature of PR is fix/refactor, let's try to get it right so that we do not have to come here again for another refactor.

c-p-b commented on frappe/desk/doctype/bulk_update/bulk_update.py:
```diff
@@ -22,7 +22,7 @@ def update(doctype, field, value, condition='', limit=500):
 		frappe.throw(_('; not allowed in condition'))
 
 	docnames = frappe.db.sql_list(
-		'''select name from `tab{0}`{1} limit 0, {2}'''.format(doctype, condition, limit)
```
Here it is not trivially easy to migrate to qb due to the condition being dynamically supplied as a string (both field name and value). I noticed this is used by client code - is it acceptable to leave this as-is?

ankush replied:
qb can accept dynamic fields like this `table[fieldname]`

c-p-b replied:
Makes sense but the variable in this case is something like "field > 5" - but it looks like Pypika allows this using Criterion.any. Will attempt to convert that way

c-p-b replied:
Criterion.any does not appear to work when an entire string is supplied - only when a pypika field is supplied with some constant value. E.g. this is supported Criterion.any(table[fieldname] == 'value') but this is not: Criterion.any(""" 'fieldname' = 'value' """)

Therefore I don't really see a way to refactor this locally without parsing the string (which of course is probably a very error prone approach and I would much prefer not to do that). Alternatively another solution would be to change the method signature which is probably also not preferred. Please advise if qb can be used here.

c-p-b commented on frappe/utils/make_random.py:
```diff
@@ -36,8 +36,13 @@ def get_random(doctype, filters=None, doc=False):
 
-	out = frappe.db.sql("""select name from `tab%s` %s
-		order by RAND() limit 0,1""" % (doctype, condition))
+	out = frappe.db.multisql({
```
This query dynamically builds condition and is not trivial to replace - ok to leave as-is?

c-p-b replied:
I refactored this but IMO it's longer and uglier than the raw SQL since you have to call DB dependent syntax in both cases. I can revert if desired.

c-p-b replied:
Criterion.any does not appear to work when an entire string is supplied - only when a pypika field is supplied with some constant value. E.g. this is supported Criterion.any(table[fieldname] == 'value') but this is not: Criterion.any(""" 'fieldname' = 'value' """). It is possible to take the inputs and parse them into the format that Criterion.any() likes in this specific case (since we have the field name and value), but it will make it even bigger/uglier than it already is when using qb. But I can go ahead and do that if we want to - it seems to make a lot more sense to just leave it as is though.

ankush replied:
@surajshetty3416  IMO we can leave queries like these out as long as they are covered 😅

c-p-b commented:
Few sider issues and some unit tests remain - looking into those today!

c-p-b commented:
Tests and sider now green!

gavindsouza commented on frappe/core/doctype/user_type/user_type.py:
```diff
@@ -36,16 +36,19 @@ def set_modules(self):
+					.select(DocType.module)
+					.where(DocType.name.isin(document_types))
+					.groupby(DocType.module)
+					.distinct()).run()
```
You can pass `pluck=True` to make ~`row[0]`~ => `row`

c-p-b replied:
👍

gavindsouza commented on frappe/utils/nestedset.py:
```diff
@@ -148,13 +151,12 @@ def rebuild_tree(doctype, parent_field):
 			(column == "") | (column.isnull())
 		)
-		.orderby(table.name, order=Order.asc)
+		.orderby(Replace(table.name, '_', ''), order=Order.asc)
```
I'm not sure what this achieves...care to elaborate? _(currently ordered by `tab{DocType`.`name`)_

c-p-b replied:
Postgres sorts underscores differently than maria due to collation differences. Anything with underscore in the front gives different results. This normalizes that result.

gavindsouza replied:
We could normalize the result at a lower level - perhaps in `frappe.qb`, `DatabaseQuery`, or somewhere similar.

cc: @saxenabhishek @Aradhya-Tripathi @surajshetty3416

cpdeethree replied:
FWIW, it's not just underscores that this happens with. Dashes, for example, are treated differently as well. This just happens to be the only place in Frappe that it happened - but in ERPNext PR there is at least one other place where dash sorting also mattered

gavindsouza replied:
@cpdeethree Did you add the `Replace` bits to fix a failing test? Because if so, I say we rather fix the tests instead. 

---

Normalizing the result like this for random queries will lead to inconsistencies in the system which will make debugging harder. Also, seems more expensive...we could try recommending similar database configurations. This shouldn't be fixed in the application IMO.

c-p-b replied:
Alright, I can back this out and use a different set of companies for this specific test suite

gavindsouza commented on frappe/email/doctype/email_queue/email_queue.py:
```diff
@@ -475,28 +475,22 @@ def get_unsubscribed_user_emails(self):
+										) | EmailUnsubscribe.global_unsubscribe == 1
+									)
+								).distinct()
+						).run()
+							
+		unsubscribed = list(unsubscribed[0]) if unsubscribed else []
```
```suggestion
						).run(pluck=True)
```

gavindsouza commented on frappe/desk/treeview.py:
```diff
@@ -40,22 +43,15 @@ def get_children(doctype, parent='', **filters):
+	meta = frappe.get_meta(doctype)
+	title_field = meta.get('title_field') or 'name'
+	title = Field(title_field)
+	return (frappe.qb.from_(DocType)
+		.select(DocType.name.as_('value'), title.as_('title'), DocType.is_group.as_('expandable'))
+		.where((Coalesce(parent_field, '') == parent) & (DocType.docstatus < 2))).run()
```
This change renders the `ignore_permissions` kwarg useless. I think we need to revert this back to `frappe.get_list` for the permission checks.

c-p-b replied:
👍

c-p-b replied:
Now reverted back to get_list

gavindsouza commented on frappe/desk/page/setup_wizard/setup_wizard.py:
```diff
@@ -405,6 +404,7 @@ def make_records(records, debug=False):
 
 		try:
 			doc.insert(ignore_permissions=True)
+			frappe.db.commit()
```
If there's a commit here, there should also be a rollback to revert every errored Document creation

c-p-b replied:
👍

gavindsouza commented on frappe/model/base_document.py:
```diff
@@ -345,7 +345,7 @@ def get_parentfield_of_doctype(self, doctype):
 		return fieldname[0] if fieldname else None
 
-	def db_insert(self):
+	def db_insert(self, db_auto_commit = False):
```
I don't think this is a good idea. This encourages poor design choices, and I can see how this would lead to data inconsistencies.

c-p-b replied:
If we can force the commits to happen upstream appropriately, this can be removed. I just could not figure out a way to do it due to the way the code is setup higher in the stack. To reproduce you can check out the upstream pr and remove the calls where I have set this to true and try running the test suite and see the failures - the ones in india_setup.py were particularly problematic

gavindsouza commented on frappe/model/document.py:
```diff
@@ -198,7 +198,7 @@ def raise_no_permission_to(self, perm_type):
 
 	def insert(self, ignore_permissions=None, ignore_links=None, ignore_if_duplicate=False,
-				ignore_mandatory=None, set_name=None, set_child_names=True):
+				ignore_mandatory=None, set_name=None, set_child_names=True, db_auto_commit = False):
```
Bad idea to commit mid Document insert.

c-p-b replied:
Absolutely - however it is the only way to get some tests to function properly and this was the lowest touch way I could manage to do it. If there is a better way I am 100% open to it.

c-p-b replied:
This is now reverted to its original state - no more autocommit

gavindsouza commented on frappe/custom/doctype/property_setter/property_setter.py:
```diff
@@ -88,7 +88,7 @@ def make_property_setter(doctype, fieldname, property, value, property_type, for
 	property_setter.flags.ignore_permissions = True
 	property_setter.flags.validate_fields_for_doctype = validate_fields_for_doctype
-	property_setter.insert()
+	property_setter.insert(db_auto_commit = frappe.flags.in_test or frappe.flags.in_setup or frappe.flags.in_install)
```
I'd suggest we find a previous step in the setup wizard or install process which can be described as a single step and keep it atomic instead.

Or increase the `MAX_WRITES_PER_TRANSACTION` for the DB connection in the respective setup or install instance.

c-p-b replied:
If you can somehow get this to work upstream in the ERPNext PR I am all ears, I tried several different things before resulting to this including forcing a commit directly after this method is called in india_setup.py with frappe.db.commit() (which did not work). The problem where this specific function is used is primarily in india_setup.py and basically causes a bunch of tests to fail for the 16 character document name length requirement for India. This property setter function sets the docfield rule that changes the naming rule for Indian countries. In that case Maria reads the new rule and creates names with the 16 character length appropriately but postgres does not unless you force commit the data here.

gavindsouza replied:
Have you tried increasing `frappe.db.MAX_WRITES_PER_TRANSACTION` (or setting it to `math.inf`) and committing post execution for the context of `make_property_setters`? (in https://github.com/frappe/erpnext/blob/3607f51610f4700acb1fd763dee8e2a0082a4a37/erpnext/regional/india/setup.py)

c-p-b replied:
I will try this today and report back

c-p-b replied:
@gavindsouza I did as you recommended and it appeared to work:

change: https://github.com/frappe/erpnext/pull/28359/commits/d0a6b9b3b163da878198de26ef62594901b9038c
clean maria: https://github.com/frappe/erpnext/runs/4653436143?check_suite_focus=true
clean postgres (minus one flaky unrelated test): https://github.com/frappe/erpnext/runs/4653436083?check_suite_focus=true


I had also called this autocommit in a few other places throughout my Erpnext changes so I am going to back out of those as well and assuming there are no further issues I will go ahead and revert all of these auto_commit changes

c-p-b replied:
This is now reverted back to its original state - no more autocommit

gavindsouza commented on frappe/database/postgres/database.py:
```diff
@@ -73,13 +73,22 @@ def get_connection(self):
 		return conn
 
-	def escape(self, s, percent=True):
+	def escape(self, s, percent=True, underscore=False):
```
I think the docstring could be fixed too ;)

c-p-b replied:
👍

c-p-b replied:
Docstring updated 👍

gavindsouza commented on frappe/custom/doctype/custom_field/custom_field.py:
```diff
@@ -135,7 +135,7 @@ def create_custom_field(doctype, df, ignore_validate=False):
 		custom_field.update(df)
 		custom_field.flags.ignore_validate = ignore_validate
-		custom_field.insert()
+		custom_field.insert(db_auto_commit = frappe.flags.in_test or frappe.flags.in_install or frappe.flags.in_setup_wizard)
```
Same comment as I've added in `property_setter.py`

c-p-b replied:
This is now removed

gavindsouza commented on frappe/model/db_query.py:
```diff
@@ -128,6 +128,20 @@ def build_and_run(self):
 			args.fields = 'distinct ' + args.fields
 			args.order_by = '' # TODO: recheck for alternative
 
+		if args.order_by and args.group_by:
```
What does this block aim to do? Can we have some comments added here?

c-p-b replied:
Postgres requires anything that appears in the select fields to also appear in the order by and group by statements. MariaDB does not. This addresses that issue. I can add a comment to that effect.

c-p-b replied:
Comment added

gavindsouza commented:
Also, to fix the "Semantic Pull Request" check, we'd have to rebase and reword the commit messages in the first 5 commits by @Verequies

c-p-b commented:
Most open items have been addressed - one still open (nestedset.py) and rest back for review.

Regarding the semantic PR - I don't really want to touch @Verequies commits and have no idea if he is around right now - is it possible to do a squash and merge instead? I'm happy to reorganize my commits however makes sense.

gavindsouza commented on frappe/database/postgres/database.py:
```diff
@@ -73,13 +73,22 @@ def get_connection(self):
 			s = s.replace("%", "%%")
+		
+		if underscore:
+			s = s.replace("_", "//_")
```
Wondering if you meant `\\_` instead. (Just a hunch)

<!-- https://stackoverflow.com/a/38084865/10309266 -->

c-p-b replied:
I will also remove this - it was actually only used to resolve a single issue in company setup which is no longer happening due to other commits.

gavindsouza commented on frappe/desk/page/setup_wizard/setup_wizard.py:
```diff
@@ -417,6 +417,7 @@ def make_records(records, debug=False):
 				raise
 
 		except Exception as e:
+			frappe.db.rollback()
```
Can we add this in the block that _(doesn't)_ handle DuplicateEntryError too?

c-p-b replied:
👍

c-p-b replied:
So I just tried adding a rollback when DuplicateEntryError happens and I guess some silent duplicateentryerrors happen during maria bootstrap because I got a bunch of test failures https://github.com/frappe/erpnext/runs/4661424145?check_suite_focus=true

After I reverted that change the failures went away

gavindsouza commented on frappe/model/base_document.py:
```diff
@@ -564,10 +564,12 @@ def get_msg(df, docname):
 						# cache a single value type
-						values = frappe._dict(name=frappe.db.get_value(doctype, docname,
-							'name', cache=True))
+						values = frappe.db.get_value(doctype, docname,
+							'name', cache=True, as_dict = True)
+						if not values:
+							values = frappe._dict(name=None)
```
This change seems quite unnecessary. Can we revert this?

c-p-b replied:
In my early dev work it was actually causing LinkValidationErrors in the old form - but I just tested now and it seems one of my other commits fixed that issue, so this can be reverted.

gavindsouza commented on frappe/model/base_document.py:
```diff
@@ -564,10 +564,12 @@ def get_msg(df, docname):
+							values = frappe._dict(name=None)
 					else:
-						values_to_fetch = ['name'] + [_df.fetch_from.split('.')[-1]
+						values_to_fetch = ['name'] + [_df.fetch_from.split('.')[-1].lower() # lowercase column for postgres
```
Is lowercase necessary?

<img width="454" alt="Screenshot 2021-12-29 at 4 42 56 PM" src="https://user-images.githubusercontent.com/36654812/147656586-24d8690a-2e4c-453a-a46d-936c80dbbf80.png">

c-p-b replied:
Same as above, this was apparently fixed by another change and can now be reverted

gavindsouza commented on frappe/model/base_document.py:
```diff
@@ -601,7 +603,7 @@ def get_msg(df, docname):
 
 	def set_fetch_from_value(self, doctype, df, values):
-		fetch_from_fieldname = df.fetch_from.split('.')[-1]
+		fetch_from_fieldname = df.fetch_from.split('.')[-1].lower()
```
Same concern as one for L572

gavindsouza commented on frappe/model/base_document.py:
```diff
@@ -771,7 +773,8 @@ def _validate_update_after_submit(self):
 					self_value = self.get_value(key)
-
+				#postgres stores values as datetime.time, maria as timedelta
+				db_value = datetime.timedelta(hours=db_value.hour, minutes = db_value.minute, seconds = db_value.second, microseconds = db_value.microsecond) if isinstance(self_value, datetime.timedelta) and isinstance(db_value, datetime.time) else db_value
```
Style suggestion for readability:

```suggestion
				# Postgres stores values as `datetime.time`, MariaDB as `timedelta`
				if isinstance(self_value, datetime.timedelta) and isinstance(db_value, datetime.time):
					db_value = datetime.timedelta(hours=db_value.hour, minutes=db_value.minute, seconds=db_value.second, microseconds=db_value.microsecond)
```

c-p-b replied:
👍

gavindsouza commented on frappe/model/db_query.py:
```diff
@@ -127,6 +127,22 @@ def build_and_run(self):
 			args.fields = 'distinct ' + args.fields
 			args.order_by = '' # TODO: recheck for alternative
+			
+		# Postgres requires any field that appears in the select clause to also 
+		# appear in the order by and group by clause
+		if args.order_by and args.group_by:
+			order_field = args.order_by
+
+			for r in (" order by ", " asc", " ASC", " desc", " DESC"):
+				order_field = order_field.replace(r, "")
+
+			if order_field not in args.fields:
+				order_fieldm = order_field.replace("`", "")
+				if "." in order_fieldm:
+					args.fields += ", MAX(" + order_fieldm.split(".")[1] + ") as `" + order_fieldm + "`"
+				else:
+					args.fields += ", MAX(" + order_fieldm + ") as `" + order_fieldm + "`"
+				args.order_by = args.order_by.replace(order_field, "`" + order_fieldm + "`")
```
This should probably be moved in a separate method, perhaps under `prepare_select_args` and have this logic executed only if db_type is postgres...given it's unnecessary processing for MariaDB

c-p-b replied:
Done

gavindsouza commented on frappe/model/db_query.py:
```diff
@@ -127,6 +127,22 @@ def build_and_run(self):
+				order_field = order_field.replace(r, "")
+
+			if order_field not in args.fields:
+				order_fieldm = order_field.replace("`", "")
+				if "." in order_fieldm:
+					args.fields += ", MAX(" + order_fieldm.split(".")[1] + ") as `" + order_fieldm + "`"
+				else:
+					args.fields += ", MAX(" + order_fieldm + ") as `" + order_fieldm + "`"
```
```suggestion
				order_fieldm = order_field.replace("`", '"')
				args.fields += f", MAX({order_fieldm}) as `{order_fieldm}`"
```

We don't need to separate the column from the table so I think this works too. Although I don't wholly understand the reason for adding `MAX(...)` here.

---

Also, I'd imagine we'd do some post-processing to remove the extra fields. This step seems unnecessary for results of type `List[Dict]` but could be problematic _(& possibly breaking)_ while `as_list=True`.

c-p-b replied:
So I already incorporated the changes here that you suggested (it just doesn't show up here as an updated diff because they are combined with another change). Do we need to include said post-processing to remove the extra fields as part of this PR? If so, where should I include that?

c-p-b replied:
After testing the more concise format I discovered that it was necessary to actually separate the column to the table. The reason being is because `tabTable.field` needs to be converted `tabTable`.`field` for this to work properly so we do need to parse the string. 

Also, I noticed that I did not answer your original question regarding MAX(). The  reason the MAX() is necessary is because in some circumstances frappe inserts docstatus and modified fields into the results of sql calls even if you did not specificially request them. That combined with the requirement of aggregate function for postgres means that those fields must be included as an aggregate function. MAX was chosen here because it just gives the most recent/highest of these fields. An example of where this happens upstream is the function get_dashboard_info

TurkerTunali commented on frappe/core/doctype/user_type/user_type.py:
```diff
@@ -36,16 +36,19 @@ def set_modules(self):
+		DocType = frappe.qb.DocType("DocType")
+
+		document_types = [d.document_type for d in self.user_doctypes] or ['']
+		modules = (frappe.qb.from_(DocType)
```
I think we need to refactor this in the frappe.get_all() method. So that all frappe.get* parameters continue to work as before.

gavindsouza replied:
@TurkerTunali  Can you specify your concerns with this query conversion?

TurkerTunali replied:
Here we are getting modules. I am not sure but instead of refoctoring here by eliminating the "frappe.get_all()" and using "frappe.qb.DocType()" we should refactor this in "frappe.get_all()" method only. 

Most probably we are using frappe.get_all() everywhere and it will require significant effort to refactor everywhere.

c-p-b committed:
fix: refactor postgres arg prep into its own function

c-p-b committed:
fix: Revert changes made to fix collation differences

Due to collation differences in MariaDB and Postgres, tests gave
inconsistent results. This was to be handled in tests alone instead of
the application. It was unnecessary. Collation changes should be made at DBMS config
level only. Accomodating for those in the application will unnecessarily
degrade performance for everyone.

Other changes:
* use pluck in user_type
* revert ordering in nestedset
* revert parsing in order_field
* use preferred APIs & styling

Co-authored-by: gavin <gavin18d@gmail.com>

ankush commented:
@cpdeethree can you update the branch so changes from https://github.com/frappe/frappe/pull/15495 are also added.

gavindsouza commented:
At this point, we have two points of concern:

- [x] Use of Replace in the query in nestedset.py (which is yet to be reverted imo)
- [ ]  `DatabaseQuery.prepare_select_args` (this bit I'm going to need some time to study the impact and any changes if required; I'll update accordingly)

gavindsouza commented:
I've managed to group, squash and re-write commits on this branch (ref: https://github.com/frappe/frappe/pull/15553). Bought em down from 32 to 4. This will clear up a lot of wrong blames from all the reverted code and fix the semantic commits check.

@cpdeethree and @Verequies let me know if you have any concerns. Planning on force pushing on this branch.

c-p-b commented:
@gavindsouza sounds good to me

gavindsouza commented:
@cpdeethree Can you add a test to validate the changes in DatabaseQuery introduced by the `prepare_select_args` method?

gavindsouza committed:
refactor(minor): DatabaseQuery.prepare_select_args

* Used regex to replace instead of re-iteration (for-loop)
* Used f-strings for better readability
* Renamed and re-ordered vars, code block

gavindsouza committed:
Merge branch 'develop' into frappe-postgres-changes

c-p-b committed:
fix: add test_for prepare_select_args

c-p-b commented:
> @cpdeethree Can you add a test to validate the changes in DatabaseQuery introduced by the `prepare_select_args` method?

added test_prepare_select_args
